Reconfigure particle flow config files

hatakeyamak opened the pull request:
#### PR description:

Reconfigure particleFlow_cfi and make clear the parameters used for PFEGammaFilters, PFMuonAlgo, i.e. different components of PFAlgo. Also used this opportunity to remove obsolete config parameters.

This is a follow-up of
https://github.com/cms-sw/cmssw/pull/27782
and fillDescriptions are introduced, and reduced ParameterSets use a little bit wrt PR27782

#### PR validation:

Compiles and run a few matrix tests with success (scenarios for 2017, 2018, and phase2)

#### if this PR is a backport please specify the original PR:

This is not a backport PR.

@bendavid @jpata @guitargeek

hatakeyamak committed:
Clean up the particleFlow config file and make associated changes

hatakeyamak committed:
Clean up the particleFlow config file and make associated changes

hatakeyamak committed:
Clean up the particleFlow config file and make associated changes

hatakeyamak committed:
Clean up the particleFlow config file and make associated changes

hatakeyamak committed:
code-check

hatakeyamak committed:
code-check

hatakeyamak committed:
code-check

hatakeyamak committed:
code-check

hatakeyamak committed:
code-format

hatakeyamak committed:
code-format

hatakeyamak committed:
code-format

hatakeyamak committed:
code-format

hatakeyamak committed:
introduce fillDescriptions

hatakeyamak committed:
introduce fillDescriptions

hatakeyamak committed:
solve conflicts

hatakeyamak committed:
solve conflicts

hatakeyamak committed:
solve conflicts

hatakeyamak committed:
cleanup. reduce parametersets usage somewhat.

hatakeyamak committed:
cleanup. reduce parametersets usage somewhat.

cmsbuild commented:
The code-checks are being triggered in jenkins.

hatakeyamak commented:
Actually, this will fail the jenkins tests that runs HLT. I am not familiar with what to do to update HLTrigger/Configuration/python/HLT_GRun_cff.py reflecting the changes here, and suggestions will be appreciated.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28110/12125

- This PR adds an extra 208KB to repository

cmsbuild commented:
A new Pull Request was created by  @hatakeyamak (Kenichi Hatakeyama) for master.

It involves the following packages:

RecoParticleFlow/PFProducer


@perrotta, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@mmarionncern, @lgray, @seemasharmafnal, @bachtis, @cbernet this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

Martin-Grunewald commented on RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporter.cc:
```diff
@@ -21,7 +21,9 @@ class GeneralTracksImporter : public BlockElementImporterBase {
-    pfmu_ = std::unique_ptr<PFMuonAlgo>(new PFMuonAlgo(conf));
+    bool postMuonCleaning =
+        conf.existsAs<bool>("postMuonCleaning") ? conf.getParameter<bool>("postMuonCleaning") : false;
+    pfmu_ = std::unique_ptr<PFMuonAlgo>(new PFMuonAlgo(conf, postMuonCleaning));
```
Why is this existsAs needed here (and in some other locations) if the parameter is part of fillDescriptions ayway?

hatakeyamak replied:
This comes from PFBlockProducer.cc (not PFProducer.cc) for which we haven't introduced fillDescriptions yet...

cmsbuild commented:
-1

Tested at: 5ea0bf620a67850a004fc54c5cc7a97523e035b5

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2f6d58/2776/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals  AddOn**

* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
140.56 step2
<pre>runTheMatrix-results/140.56_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18/step2_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18.log

</pre>158.0 step2
<pre>runTheMatrix-results/158.0_HydjetQ_B12_5020GeV_2018_ppReco+HydjetQ_B12_5020GeV_2018_ppReco+DIGIHI2018PPRECO+RECOHI2018PPRECO+ALCARECOHI2018PPRECO+HARVESTHI2018PPRECO/step2_HydjetQ_B12_5020GeV_2018_ppReco+HydjetQ_B12_5020GeV_2018_ppReco+DIGIHI2018PPRECO+RECOHI2018PPRECO+ALCARECOHI2018PPRECO+HARVESTHI2018PPRECO.log

</pre>11634.0 step2
<pre>runTheMatrix-results/11634.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2021_GenSimFull+DigiFull_2021+RecoFull_2021+HARVESTFull_2021+ALCAFull_2021/step2_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2021_GenSimFull+DigiFull_2021+RecoFull_2021+HARVESTFull_2021+ALCAFull_2021.log

</pre>12434.0 step2
<pre>runTheMatrix-results/12434.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023_GenSimFull+DigiFull_2023+RecoFull_2023+HARVESTFull_2023+ALCAFull_2023/step2_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023_GenSimFull+DigiFull_2023+RecoFull_2023+HARVESTFull_2023+ALCAFull_2023.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsRun /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_0_X_2019-10-02-2300/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Thu Oct  3 17:24:35 2019-date Thu Oct  3 17:23:23 2019 s - exit: 18688
cmsDriver.py RelVal -s HLT:HIon,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_HIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2018_pp_on_AA --processName=HLTRECO --filein file:RelVal_Raw_HIon_DATA.root --fileout file:RelVal_Raw_HIon_DATA_HLT_RECO.root : FAILED - time: date Thu Oct  3 17:24:35 2019-date Thu Oct  3 17:23:23 2019 s - exit: 256
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_0_X_2019-10-02-2300/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Thu Oct  3 17:30:20 2019-date Thu Oct  3 17:23:26 2019 s - exit: 18688
cmsDriver.py RelVal -s HLT:GRun,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_GRun --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run3 --processName=HLTRECO --filein file:RelVal_Raw_GRun_MC.root --fileout file:RelVal_Raw_GRun_MC_HLT_RECO.root : FAILED - time: date Thu Oct  3 17:30:20 2019-date Thu Oct  3 17:23:26 2019 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_0_X_2019-10-02-2300/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Thu Oct  3 17:29:17 2019-date Thu Oct  3 17:23:48 2019 s - exit: 18688
cmsDriver.py RelVal -s HLT:HIon,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_HIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2018_pp_on_AA --processName=HLTRECO --filein file:RelVal_Raw_HIon_MC.root --fileout file:RelVal_Raw_HIon_MC_HLT_RECO.root : FAILED - time: date Thu Oct  3 17:29:17 2019-date Thu Oct  3 17:23:48 2019 s - exit: 256
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_0_X_2019-10-02-2300/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Thu Oct  3 17:29:24 2019-date Thu Oct  3 17:24:02 2019 s - exit: 18688
cmsDriver.py RelVal -s HLT:GRun,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_GRun --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2018 --processName=HLTRECO --filein file:RelVal_Raw_GRun_DATA.root --fileout file:RelVal_Raw_GRun_DATA_HLT_RECO.root : FAILED - time: date Thu Oct  3 17:29:24 2019-date Thu Oct  3 17:24:02 2019 s - exit: 18688

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

hatakeyamak commented:
As I wrote earlier, I wasn't sure what to update HLT configs, so these failures were sort of expected. I will appreciate advice on this.

Dr15Jones commented:
> As I wrote earlier, I wasn't sure what to update HLT configs, so these failures were sort of expected. I will appreciate advice on this.

If the HLT module uses a different module label, it is possible to have different `fillDescriptions` descriptions for different module labels.

Martin-Grunewald commented:
Is `bCalibSecondary` gone for good?
In that case put a customisation into HLTrigger/Configuration/python/customizeHLTforCMSSW.py
to kill this parameter in any instance of the producer. This would be temporary until the HLT menu in
ConfDB is updated with your new cfi from fillDescriptions. Newly added parameters should be already taken care of fillDescriptions. In any case it would be good if there is only ONE set of top-level py parameters for the producer, as this is what ConfDB can handle.

hatakeyamak commented:
Hi Martin. Thank you. Yes, bCalibSecondary should be gone for good (as far as I can tell, it was not used for quite some time). There are several others that are removed because they were not used or moved into sub-psets.

I am updating HLTrigger/Configuration/python/customizeHLTforCMSSW.py, following your suggestions.

hatakeyamak committed:
kill unused or moved parameters via HLTrigger/Configuration/python/customizeHLTforCMSSW.py

hatakeyamak committed:
kill unused or moved parameters via HLTrigger/Configuration/python/customizeHLTforCMSSW.py

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28110/12189

- This PR adds an extra 220KB to repository

cmsbuild commented:
Pull request #28110 was updated. @perrotta, @Martin-Grunewald, @cmsbuild, @fwyzard, @slava77 can you please check and sign again.

hatakeyamak commented:
Wait. Need to check one more thing.

hatakeyamak committed:
reorganize config files for PFProducer/particleFlow

hatakeyamak committed:
reorganize config files for PFProducer/particleFlow

hatakeyamak committed:
reorganize config files for PFProducer/particleFlow

hatakeyamak committed:
reorganize config files for PFProducer/particleFlow

cmsbuild commented:
The code-checks are being triggered in jenkins.

hatakeyamak commented:
Updated  HLTrigger/Configuration/python/customizeHLTforCMSSW.py based on comments from @Martin-Grunewald and made some cleanups.

Tested with matrix tests - 10024.0, 10824.0, 20434.0, 11634.0, 140.56, 158.0 and looks ok to me.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28110/12193

- This PR adds an extra 220KB to repository

cmsbuild commented:
Pull request #28110 was updated. @perrotta, @Martin-Grunewald, @cmsbuild, @fwyzard, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: a57622a9079dbe12401985dca5e00d75a60d8d22

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2f6d58/2867/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
1001.0 step5
<pre>runTheMatrix-results/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVDSIPIXELCALRUN1+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5/step5_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVDSIPIXELCALRUN1+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5.log

</pre>1330.0 step5
<pre>runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15_L1TMuDQM+HARVESTUP15_L1TMuDQM+NANOUP15/step5_ZMM_13+ZMM_13+DIGIUP15+RECOUP15_L1TMuDQM+HARVESTUP15_L1TMuDQM+NANOUP15.log

</pre>20034.0 step1
<pre>runTheMatrix-results/20034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D35_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D35+RecoFullGlobal_2026D35+HARVESTFullGlobal_2026D35/step1_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D35_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D35+RecoFullGlobal_2026D35+HARVESTFullGlobal_2026D35.log

</pre>158.0 step5
<pre>runTheMatrix-results/158.0_HydjetQ_B12_5020GeV_2018_ppReco+HydjetQ_B12_5020GeV_2018_ppReco+DIGIHI2018PPRECO+RECOHI2018PPRECO+ALCARECOHI2018PPRECO+HARVESTHI2018PPRECO/step5_HydjetQ_B12_5020GeV_2018_ppReco+HydjetQ_B12_5020GeV_2018_ppReco+DIGIHI2018PPRECO+RECOHI2018PPRECO+ALCARECOHI2018PPRECO+HARVESTHI2018PPRECO.log

</pre>25202.0 step3
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25+NANOUP15_PU25/step3_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25+NANOUP15_PU25.log

</pre>1000.0 step4
<pre>runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step4_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

</pre>10042.0 step3
<pre>runTheMatrix-results/10042.0_ZMM_13+ZMM_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017+ALCAFull_2017+NanoFull_2017/step3_ZMM_13+ZMM_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017+ALCAFull_2017+NanoFull_2017.log

</pre>10224.0 step2
<pre>runTheMatrix-results/10224.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017PU_GenSimFull+DigiFullPU_2017PU+RecoFullPU_2017PU+HARVESTFullPU_2017PU+NanoFull_2017PU/step2_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017PU_GenSimFull+DigiFullPU_2017PU+RecoFullPU_2017PU+HARVESTFullPU_2017PU+NanoFull_2017PU.log

</pre>10024.0 step3
<pre>runTheMatrix-results/10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017+ALCAFull_2017+NanoFull_2017/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017+ALCAFull_2017+NanoFull_2017.log

</pre>10824.0 step2
<pre>runTheMatrix-results/10824.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+HARVESTFull_2018+ALCAFull_2018+NanoFull_2018/step2_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+HARVESTFull_2018+ALCAFull_2018+NanoFull_2018.log

</pre>11634.0 step2
<pre>runTheMatrix-results/11634.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2021_GenSimFull+DigiFull_2021+RecoFull_2021+HARVESTFull_2021+ALCAFull_2021/step2_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2021_GenSimFull+DigiFull_2021+RecoFull_2021+HARVESTFull_2021+ALCAFull_2021.log

</pre>12434.0 step2
<pre>runTheMatrix-results/12434.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023_GenSimFull+DigiFull_2023+RecoFull_2023+HARVESTFull_2023+ALCAFull_2023/step2_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023_GenSimFull+DigiFull_2023+RecoFull_2023+HARVESTFull_2023+ALCAFull_2023.log

</pre>21234.0 step1
<pre>runTheMatrix-results/21234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D44_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D44+RecoFullGlobal_2026D44+HARVESTFullGlobal_2026D44/step1_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D44_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D44+RecoFullGlobal_2026D44+HARVESTFullGlobal_2026D44.log

</pre>136.85 step4
<pre>runTheMatrix-results/136.85_RunEGamma2018A+RunEGamma2018A+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Offline_L1TEgDQM+HARVEST2018_L1TEgDQM/step4_RunEGamma2018A+RunEGamma2018A+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Offline_L1TEgDQM+HARVEST2018_L1TEgDQM.log

</pre>22034.0 step1
<pre>runTheMatrix-results/22034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D46_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D46+RecoFullGlobal_2026D46+HARVESTFullGlobal_2026D46/step1_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D46_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D46+RecoFullGlobal_2026D46+HARVESTFullGlobal_2026D46.log

</pre>140.56 step2
<pre>runTheMatrix-results/140.56_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18/step2_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18.log

</pre>20434.0 step1
<pre>runTheMatrix-results/20434.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D41_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D41+RecoFullGlobal_2026D41+HARVESTFullGlobal_2026D41/step1_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D41_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D41+RecoFullGlobal_2026D41+HARVESTFullGlobal_2026D41.log

</pre>250202.181 step2
<pre>runTheMatrix-results/250202.181_TTbar_13UP18+TTbar_13UP18+PREMIXUP18_PU25+DIGIPRMXLOCALUP18_PU25+RECOPRMXUP18_PU25+HARVESTUP18_PU25/step2_TTbar_13UP18+TTbar_13UP18+PREMIXUP18_PU25+DIGIPRMXLOCALUP18_PU25+RECOPRMXUP18_PU25+HARVESTUP18_PU25.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

hatakeyamak commented:
The failures might be due to some database access issue rahter than problem of this PR. I re-ran some failed tests locally, and they ran ok. Can the tests be re-triggered?

perrotta commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

hatakeyamak commented:
Thank you @perrotta

cmsbuild commented:
+1
Tested at: a57622a9079dbe12401985dca5e00d75a60d8d22
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2f6d58/2871/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2f6d58/2871/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2961056
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2960714
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

hatakeyamak commented:
Just checking in to see how this PR looks for reviewers.

fabiocos commented:
@Martin-Grunewald @perrotta @slava77 latest test was technically positive, could you please have a look at this PR?

Martin-Grunewald commented:
+1

slava77 commented on RecoParticleFlow/PFProducer/plugins/PFProducer.cc:
```diff
@@ -333,3 +309,233 @@ void PFProducer::produce(Event& iEvent, const EventSetup& iSetup) {
+
+  // For PFEGammaFilters
+  {
+    edm::ParameterSetDescription psd_PFEgammaFilters;
```
it looks like this PSet  description fully belongs to the PFEgammaFilters (not used anywhere else by the PFProducer). This is a pattern usually implies that the defaults better be defined by ```PFEgammaFilters``` 

A suggested approach for this for a general case is https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#A_Plugin_Module_Using_Another_He
Here, since the factory pattern is not used for the helper class, a simpler solution is to define in ```PFEgammaFilters```
```c++
static void
fillPSetDescription(edm::ParameterSetDescription& iDesc) {
    iDesc.add<double>("electron_iso_pt", 10.0);
 ...
}
```

hatakeyamak replied:
I have to admit I am still not sure how to do this and need help.
We don't really use "factory", so I am not sure how this recipe will fit it into this use case.

Some specific questions are:
+ PFEGammaFilters may be considered as a helper, but it's not a plugin module right now. We need to make it as a plugin?
+ Do we need to introduce a new "factory" that creates PFEGammaFilters as a new plugin?
I think I need a little more guidance to get this implemented.

hatakeyamak replied:
Maybe a little more in detail, by looking at:
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#A_Plugin_Module_Using_Another_He
I probably understand steps 1,2,3, but steps 4,5,6 are unclear (which files these lines need to go in?) to me right now.

slava77 replied:
the link to the twiki was given for a more complete reference for a general case.
For this specific case, please follow the suggested code snippets:

add in PFEgammaFilters.h,cc: (as already posted above)
```c++
static void PFEgammaFilters::fillPSetDescription(edm::ParameterSetDescription& iDesc) {
    iDesc.add<double>("electron_iso_pt", 10.0);
 ...
}
```
then in this file replace L349-L409 with 
```c++
PFEgammaFilters::fillPSetDescription(psd_PFEgammaFilters);
```

slava77 commented on RecoParticleFlow/PFProducer/plugins/PFProducer.cc:
```diff
@@ -333,3 +309,233 @@ void PFProducer::produce(Event& iEvent, const EventSetup& iSetup) {
+    psd_PFMuonAlgo.add<double>("punchThroughFactor", 3.0);
+    psd_PFMuonAlgo.add<double>("punchThroughMETFactor", 4.0);
+    psd_PFMuonAlgo.add<double>("cosmicRejectionDistance", 1.0);
+    desc.add<edm::ParameterSetDescription>("PFMuonAlgoParameters", psd_PFMuonAlgo);
```
similarly here, the defaults can be defined in ```PFMuonAlgo::fillPSetDescription```

slava77 commented on RecoParticleFlow/PFProducer/plugins/PFProducer.cc:
```diff
@@ -333,3 +309,233 @@ void PFProducer::produce(Event& iEvent, const EventSetup& iSetup) {
+      "goodPixelTrackDeadHcal_maxLost3Hit",
+      0);  // max missing outer hits for a track with 3 valid pixel layers (can set to -1 to reject all these tracks)
+  desc.add<int>("goodPixelTrackDeadHcal_maxLost4Hit",
+                1);  // max missing outer hits for a track with >= 4 valid pixel layers
```
due to clang-format driven line breaks, it's more practical to put the comment above the line.
Anyways, these seem to fall under comments that better be known to the parameters with ```setComment```

slava77 commented on RecoParticleFlow/PFProducer/plugins/PFProducer.cc:
```diff
@@ -333,3 +309,233 @@ void PFProducer::produce(Event& iEvent, const EventSetup& iSetup) {
+  desc.add<std::vector<double>>("calibHF_a_EMonly", {1., 1., 1., 1., 1., 1., 1., 1., 1., 1.});
+  desc.add<std::vector<double>>("calibHF_a_EMHAD", {1., 1., 1., 1., 1., 1., 1., 1., 1., 1.});
+  desc.add<std::vector<double>>("calibHF_b_HADonly", {1., 1., 1., 1., 1., 1., 1., 1., 1., 1.});
+  desc.add<std::vector<double>>("calibHF_b_EMHAD", {1., 1., 1., 1., 1., 1., 1., 1., 1., 1.});
```
these are all different from what was in particleFlow_cfi.py before.
Is this intentional?
is this hardcoded calib machinery still needed?

hatakeyamak replied:
These are not, but I didn't want to completely remove them. So, parameters are just set like these.

slava77 commented on RecoParticleFlow/PFProducer/plugins/PFProducer.cc:
```diff
@@ -333,3 +309,233 @@ void PFProducer::produce(Event& iEvent, const EventSetup& iSetup) {
+  desc.add<std::vector<double>>("calibHF_b_HADonly", {1., 1., 1., 1., 1., 1., 1., 1., 1., 1.});
+  desc.add<std::vector<double>>("calibHF_b_EMHAD", {1., 1., 1., 1., 1., 1., 1., 1., 1., 1.});
+
+  descriptions.add("particleFlow", desc);
```
it looks like only _cff file should be exposed to users for a consistent use.
For this case, since there were known use cases of ```particleFlow_cfi```, it is better to avoid confusion and name this differencely, e.g. use the default label ```pfProducer``` or ```particleFlowDefault```

hatakeyamak replied:
Now, particleFlow_cfi.py is being removed. I think I can use e.g. particleFlowDefault, but I am not so sure why it's preferred.

slava77 replied:
As in the current state of the PR, the particleFlow_cfi.py is not removed, it is instead replaced by the auto-generated particleFlow_cfi.py file which will be incomplete if anyone tries to use it.
the main reason to rename this auto-generated file is to avoid possible confusion for someone who had it in their old configuration (outside of the official CMSSW code tree).

Martin-Grunewald replied:
@slava77 What do you mean here? For ConfDB parsing we'd need a cfi file with at least all top-level parameters...

slava77 replied:
@Martin-Grunewald 
I expect that you will use particleFlow_cff, the new standard reference file.
Perhaps I need to be otherwise educated on how you pick a file for parsing

Martin-Grunewald replied:
@slava77 
If there is a cfi it picks that, rather than a cff, by default. What is in the cff and not cfi?

slava77 replied:
era-related modifications are in the cff

hatakeyamak replied:
Era-dependent modifications used to be in cfi, but i moved them to cff, because now cfi is auto-generated, and I didn't know how to keep the era-related modifications in cfi which is now auto-generated.

slava77 replied:
@Martin-Grunewald 
to expand on my note in https://github.com/cms-sw/cmssw/pull/28110#discussion_r336455010: please clarify if there is a specific rule in the HLT confDB parsing which requires using the auto-generated cfi file (which is incomplete for offline use case) rather than the partly manually made cff file (which is fully usable for offline configs).
In other words, is there anything that would prevent you from using the particleFlow_cff?

Martin-Grunewald replied:
@slava77 
Currently ConfDB parsing uses cfi by default if there is one, but I can tell it to use a cff. In this case, does the cfi simply contain less parameters, with the rest added by hand elsewhere or by fillDescriptions of other plugings/code?

slava77 replied:
in this case it should be the same parameters, just the values can change depending on the era/modifier, which is defined/applied in the cff.

Is there some magic memory from the past in the cfi file discovery mechanism? If we rename this one to particleFlowDefault_cfi (as proposed above), would ConfDb still be able to find it?

Martin-Grunewald replied:
Yes, sure, it looks at any cfi it finds. Also, it is really more about the parameters themselves, not so much their precise values which may need adjustment for hlt anyway. So the name of the file could remain as is.

hatakeyamak replied:
ok. thx for clarification.

slava77 commented on RecoParticleFlow/PFProducer/src/PFAlgo.cc:
```diff
@@ -23,7 +23,9 @@ PFAlgo::PFAlgo(double nSigmaECAL,
       connector_() {
-  pfmu_ = std::make_unique<PFMuonAlgo>(pset);
+  const edm::ParameterSet pfMuonAlgoParams = pset.getParameter<edm::ParameterSet>("PFMuonAlgoParameters");
+  bool postMuonCleaning = pset.existsAs<bool>("postMuonCleaning") ? pset.getParameter<bool>("postMuonCleaning") : false;
```
isn't ```PFAlgo``` used only by the ```PFProducer```?
If so, the ```postMuonCleaning``` was already added via the fillDescriptions there

hatakeyamak replied:
Hmm. PFAlgo's fillDescriptions just has this parameter then? I am not sure I like the idea, but if it's really preferred, I think I can do it.

slava77 replied:
I thought this was the case from 
https://github.com/cms-sw/cmssw/pull/28110/files#diff-24b7d20f1967c466ee0593961eb6c237R325

hatakeyamak replied:
now i see what you mean. thank you.

slava77 reviewed (commented):
sorry for some delay in the review; I've accumulated some backlog

hatakeyamak commented:
Thanks @slava77 I will try to work on these suggestions later today or tomorrow, but I had a few questions which I posted above..

slava77 commented:
> Thanks @slava77 I will try to work on these suggestions later today or tomorrow, but I had a few questions which I posted above..

I just wanted to check if all questions on what to do were addressed.
@hatakeyamak please clarify.
Thank you.

hatakeyamak commented:
Yes, at least conceptually. I am just working on implementing suggestions now. I might ask for some help while implementing them.

christopheralanwest committed:
Use corrected PF cluster correction tags for 2018 design scenario

makortel committed:
Remove ParameterSet::Bool as unused

hatakeyamak committed:
clean up for deletion

hatakeyamak committed:
clean up for deletion

hatakeyamak committed:
remove unnecessary existsAs

hatakeyamak committed:
remove unnecessary existsAs

hatakeyamak committed:
cleanup commented-out lines

hatakeyamak committed:
cleanup commented-out lines

hatakeyamak committed:
introduce fillPSetDescription for helper classes

hatakeyamak committed:
introduce fillPSetDescription for helper classes

cmsbuild commented:
The code-checks are being triggered in jenkins.

hatakeyamak commented:
Thank you @slava77. I think I implemented all suggested changes.

I also introduced fillPSetDescription for PFCandConnector with the same spirit as we used for PFEGammaFilters and PFMuonAlgo.

In PFMuonAlgo, we cannot get rid of getParameter with default values, as it is used not only by PFProducer/PFAlgo but by others as well, but I synchronized default values between getParameter's in constructor and fillPSetDescriptions, and I think it should have no consequence to reco/HLT outputs.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28110/12471

- This PR adds an extra 80KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HLTrigger/Configuration/python/customizeHLTforCMSSW.py  modified in PR(s): #28163
  - File  RecoParticleFlow/PFProducer/plugins/PFEGammaProducer.cc  modified in PR(s): #28271
  - File  RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporter.cc  modified in PR(s): #28200
  - File  RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporterWithVeto.cc  modified in PR(s): #28200

cmsbuild commented:
Pull request #28110 was updated. @perrotta, @Martin-Grunewald, @cmsbuild, @fwyzard, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: b3fd33a6b3a2ac9c0e49c19a96555cae3d2f3809
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2f6d58/3195/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2f6d58/3195/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2961036
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2960693
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

hatakeyamak commented:
Comparisons look good to me.

Commit:
call close with the option to shutdown workers

Commit:
formatting

fabiocos commented:
@hatakeyamak there is a conflict introduced by the integration of #28200 which, looking at the code, seems simple to solve keeping your formulation (unless you want to make an update)

hatakeyamak committed:
Merge branch 'master' into Reconfigure_particleFlow_cfi

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28110/12534

- This PR adds an extra 296KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HLTrigger/Configuration/python/customizeHLTforCMSSW.py  modified in PR(s): #28163
  - File  RecoParticleFlow/PFProducer/plugins/PFEGammaProducer.cc  modified in PR(s): #28271


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28110/12534/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28110/12534/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

hatakeyamak committed:
code-format

hatakeyamak committed:
solve conflict

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28110/12540

- This PR adds an extra 212KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HLTrigger/Configuration/python/customizeHLTforCMSSW.py  modified in PR(s): #28163
  - File  RecoParticleFlow/PFProducer/plugins/PFEGammaProducer.cc  modified in PR(s): #28271

cmsbuild commented:
Pull request #28110 was updated. @perrotta, @Martin-Grunewald, @cmsbuild, @fwyzard, @slava77 can you please check and sign again.

hatakeyamak commented:
Thank you @fabiocos . I think I solved the conflict.
Does this PR look good now, reviewers?

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 67566738344e284517e8aebbd623ae32a8b271ed
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2f6d58/3254/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - #28305 @dan131riley: call close with the option to shutdown workers
 - #28294 @makortel: Remove ParameterSet::Bool as unused
 - #28263 @christopheralanwest: Use corrected PF cluster correction tags for 2018 design MC scenario

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2f6d58/3254/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2f6d58/3254/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2f6d58/3254/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2961036
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2960694
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/28110	6756673 
- code changes are in line with the PR description and the follow up review
    - I only inspected the value changes from the code diffs, additionally relying on lack of differences in the results
- jenkins tests pass and comparisons with the baseline show no differences

hatakeyamak commented:
Thanks @slava77.
Looks like this is still waiting for HLT signature. @fwyzard @Martin-Grunewald

Martin-Grunewald commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1
